src/random_bq_to_anndata.py

```diff
@@ -21,26 +24,105 @@ def get_cell_data(project, dataset, client, num_cells):
 
     # at some point, we will probably want create temp table of cell_ids and then JOIN on it
     # instead of an IN clause
-    sql = f"SELECT original_cell_id, cell_type, original_feature_id, feature_name, raw_counts AS count FROM `{project}.{dataset}.cas_cell_info` AS cell, `{project}.{dataset}.cas_feature_info` AS feature, `{project}.{dataset}.cas_raw_count_matrix` AS matrix WHERE matrix.cas_cell_index = cell.cas_cell_index AND matrix.cas_feature_index = feature.cas_feature_index AND" + in_clause
+    sql = f"SELECT matrix.cas_cell_index, original_cell_id, cell_type, matrix.cas_feature_index, original_feature_id, feature_name, raw_counts AS count FROM `{project}.{dataset}.cas_cell_info` AS cell, `{project}.{dataset}.cas_feature_info` AS feature, `{project}.{dataset}.cas_raw_count_matrix` AS matrix WHERE matrix.cas_cell_index = cell.cas_cell_index AND matrix.cas_feature_index = feature.cas_feature_index AND" + in_clause + " ORDER BY matrix.cas_cell_index, matrix.cas_feature_index"
     print(f"Getting {num_cells} random cells' data from {project}.{dataset}...")
     query = client.query(sql)
     return query.result()
 
 
-def random_bq_to_anndata(project, dataset, num_cells):
+def random_bq_to_anndata(project, dataset, num_cells, output_file_prefix):
     client = bigquery.Client(project=project)
+    # Note that this method requires that the result set returned by get_cell_data be sorted by cas_cell_index (or, could also be sorted by original_cell_id)
     cell_data = get_cell_data(project, dataset, client, num_cells)
-    adata = ad.AnnData()
+
+    column_count = 0
+    feature_to_column = {}
+    last_cell_type = None
+
+    # For representation of the sparse matrix
+    index_ptr = [0]
+    indices = []
+    data = []
+
+    # For storage of the metadata
+    cell_names = []
+    cell_types = []
+    feature_ids = []
+    feature_names = []
+
+    # Builds the sparse matrix for storing the original_cell_ids (as obs) and features (as var).
+    # From: https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html
+    # If we want to store the following sparse matrix:
+    # OBS \ VAR
+    #       ENS1 ENS2 ENS3 ENS4
+    # AAA   1    7     2   n/a
+    # AAC   0    n/a   4   n/a
+    # AAG   3    n/a   0   1
+    #
+    # we represent it as:
+    # index_ptr = [0, 3, 5, 8]
+    # indices   = [0, 1, 2, 0, 2, 0, 2, 3]
+    # data      = [1, 7, 2, 0, 4, 3, 0, 1]
+
+    last_original_cell_id = None
     for row in list(cell_data):
```

Comment:
if the return from `get_cell_data` is iterable it might be preferable to iterate that rather than force evaluation to a list